src/main/scala/org/globalforestwatch/features/FeatureRDD.scala

```diff
@@ -60,92 +71,137 @@ object FeatureRDD {
 
-          val geom = GeometryReducer.reduce(GeometryReducer.gpr)(
-            vector.Point(pt.getX, pt.getY)
-          )
+          // use implicit geometry converter
+          val geom: vector.Point = pt
 
           val pointFeatureId: FeatureId =
-            featureObj.getFeatureId(pointFeatureData)
+            FeatureIdFactory(featureType).fromUserData(pointFeatureData)
+
           vector.Feature(geom, pointFeatureId)
         case _ =>
-          throw new IllegalArgumentException("Point-in-polygon intersection must be points.")
+          throw new IllegalArgumentException(
+            "Point-in-polygon intersection must be points."
+          )
       }
 
-    // In case we implement this method for other geometry types we will have to split geometries
-    //      .flatMap { feature =>
-    //        splitGeometry(feature)
-    //      }
-
   }
 
-
   /*
     Convert polygon-polygon intersection join to feature RDD
    */
   def apply(
-             feature1Obj: Feature,
-             feature2Obj: Feature,
-             spatialRDD: SpatialRDD[GeoSparkGeometry],
+             feature1Type: String,
+             feature1Uris: NonEmptyList[String],
+             feature1Delimiter: String,
+             feature2Type: String,
+             feature2Uris: NonEmptyList[String],
+             feature2Delimiter: String,
              kwargs: Map[String, Any],
+             spark: SparkSession
            ): RDD[geotrellis.vector.Feature[Geometry, FeatureId]] = {
-    val scalaRDD =
-      org.apache.spark.api.java.JavaRDD.toRDD(spatialRDD.spatialPartitionedRDD)
-
-    scalaRDD
-      .flatMap {
-        case shp: GeoSparkPolygonal =>
-          val featureData = shp.getUserData.asInstanceOf[String].split('\t')
-          val writer: WKTWriter = new WKTWriter()
-          val wkt = writer.write(shp)
-
-          val reader: WKTReader = new WKTReader()
-          val geom = GeometryReducer.reduce(GeometryReducer.gpr)(
-            reader.read(wkt)
-          )
 
-          val feature1Data = featureData.head.drop(1).dropRight(1).split(',')
-          val feature1Id: FeatureId =
-            feature1Obj.getFeatureId(feature1Data, parsed = true)
-
-          val feature2Data = featureData.tail.head.drop(1).dropRight(1).split(',')
-          val feature2Id: FeatureId =
-            feature2Obj.getFeatureId(feature2Data, parsed = true)
+    val spatialDF: DataFrame = PolygonIntersectionDF(
+      feature1Uris,
+      feature1Type,
+      feature2Uris,
+      feature2Type,
+      spark,
+      kwargs,
+      feature2Delimiter = ","
+    )
+
+    val pairedRDD = spatialDF.rdd.map { row: Row =>
+      val featureId1: FeatureId = FeatureIdFactory(feature1Type)
+        .fromUserData(row.getAs[Row](0).toString, ",")
+      val featureId2: FeatureId = FeatureIdFactory(feature2Type)
+        .fromUserData(row.getAs[Row](1).toString, ",")
+      val geom = row.getAs[GeoSparkGeometry](2)
+
+      (CombinedFeatureId(featureId1, featureId2), geom)
+    }
```

Comment:
I somehow got geometries with userData = Null. I hence refactored the part where you extract the user data. I did not go via a SpatialRDD but a regular one.